patches/server/0951-More-DragonBattle-API.patch

```diff
@@ -0,0 +1,97 @@
++    public List<EndCrystal> getSpikeCrystals() {
++        List<EndCrystal> endCrystals = new java.util.ArrayList<>();
++        for(SpikeFeature.EndSpike endSpike : SpikeFeature.getSpikesForLevel(this.level)) {
++            endCrystals.addAll(this.level.getEntitiesOfClass(EndCrystal.class, endSpike.getTopBoundingBox()));
++        }
++        return endCrystals;
++    }
++    // Paper end
+ 
+     public void resetSpikeCrystals() {
+         for(SpikeFeature.EndSpike endSpike : SpikeFeature.getSpikesForLevel(this.level)) {
+diff --git a/src/main/java/org/bukkit/craftbukkit/boss/CraftDragonBattle.java b/src/main/java/org/bukkit/craftbukkit/boss/CraftDragonBattle.java
+index 9c1dbdffa18107924cdda5f935631c1e0f652455..f79da5c83ba5656b6fbcff5e050874a1d49fb988 100644
+--- a/src/main/java/org/bukkit/craftbukkit/boss/CraftDragonBattle.java
++++ b/src/main/java/org/bukkit/craftbukkit/boss/CraftDragonBattle.java
+@@ -97,4 +97,45 @@ public class CraftDragonBattle implements DragonBattle {
+     private DragonRespawnAnimation toNMSRespawnPhase(RespawnPhase phase) {
+         return (phase != RespawnPhase.NONE) ? DragonRespawnAnimation.values()[phase.ordinal()] : null;
+     }
++    // Paper start
++    @Override
++    public int getGatewayCount() {
++        return EndDragonFight.GATEWAY_COUNT - this.handle.gateways.size();
++    }
++
++    @Override
++    public boolean spawnNewGateway() {
++        return this.handle.spawnNewGatewayIfPossible();
++    }
++
++    @Override
++    public void spawnNewGateway(final io.papermc.paper.math.Position position) {
++        this.handle.spawnNewGateway(new net.minecraft.core.BlockPos(position.x(), position.y(), position.z()));
++    }
++
++    @Override
++    public java.util.List<org.bukkit.entity.EnderCrystal> getRespawnCrystals() {
++        if (this.handle.respawnCrystals == null) {
++            return java.util.Collections.emptyList();
++        }
++        final java.util.List<org.bukkit.entity.EnderCrystal> enderCrystals = new java.util.ArrayList<>();
++        for (final net.minecraft.world.entity.boss.enderdragon.EndCrystal endCrystal : this.handle.respawnCrystals) {
++            if (!endCrystal.isRemoved() && endCrystal.isAlive() && endCrystal.valid) {
++                enderCrystals.add(((org.bukkit.entity.EnderCrystal) endCrystal.getBukkitEntity()));
++            }
++        }
++        return java.util.Collections.unmodifiableList(enderCrystals);
++    }
++
++    @Override
++    public java.util.List<org.bukkit.entity.EnderCrystal> getHealingCrystals() {
++        final java.util.List<org.bukkit.entity.EnderCrystal> enderCrystals = new java.util.ArrayList<>();
++        for (final net.minecraft.world.entity.boss.enderdragon.EndCrystal endCrystal : this.handle.getSpikeCrystals()) {
++            if (!endCrystal.isRemoved() && endCrystal.isAlive() && endCrystal.valid) {
++                enderCrystals.add(((org.bukkit.entity.EnderCrystal) endCrystal.getBukkitEntity()));
++            }
```

Comment:
All these alive checks seem redundant?

Reply:
Vanilla doesn’t actually remove the entity from the list when it dies as far as I can tell. So it just has a list of non-existent entities. That’s there to just make sure they don’t sneak into the api.